### ISSUE TITLE
代码报错之后显存得不到释放

### ISSUE THREAD

**Issue by kaijieshi7** [Oneflow-Inc/oneflow#6967]
## Summary

1. 运行一个代码后报错，然后显存就被占满，得不到释放。我目前不知道怎么解决，只能重启。
2. 在pycharm和Ubuntu自带的终端都会爆炸

## Code to reproduce bug
待会儿帖文件以及如何运行
## 报错截图
![Screenshot from 2021-12-08 12-50-07](https://user-images.githubusercontent.com/19171201/145150307-2426a4e1-4016-4d01-9a54-83070ea4c4ac.png)

![Screenshot from 2021-12-08 11-43-52](https://user-images.githubusercontent.com/19171201/145150310-3d47dc40-24c1-4127-bbe6-e10ad0b3b12c.png)
## System Information
- What is your OneFlow installation (pip, source, dockerhub): pip
- OS: ubuntu16
- OneFlow version (run `python3 -m oneflow --doctor`): 0.6.0.dev20211205+cu102
- Python version: 3.8
- CUDA driver version: 440.64
- GPU models: 2080ti, CPU无核显
- Other info:

**Comment by chengtbf** [comment]
你试着 kill 一下僵尸进程呢？  看样子虽然是 core dump 了，但是进程并没有退出 😂

**Comment by kaijieshi7** [comment]
> 你试着 kill 一下僵尸进程呢？ 看样子虽然是 core dump 了，但是进程并没有退出 joy
用这个命令找不到僵尸进程
ps -A -o stat,ppid,pid,cmd | grep -e '^[Zz]'
![Screenshot from 2021-12-08 13-00-08](https://user-images.githubusercontent.com/19171201/145150950-e06f5649-c071-4a40-817c-c3d5d4189a31.png)

**Comment by kaijieshi7** [comment]
[train_cifar(py).txt](https://github.com/Oneflow-Inc/oneflow/files/7673477/train_cifar.py.txt)
[train_cifar(sh).txt](https://github.com/Oneflow-Inc/oneflow/files/7673479/train_cifar.sh.txt)

如何复现：

1. 将他们改为.sh和.py文件
2. 放到 https://github.com/Oneflow-Inc/models/tree/skj_gan_stylegan2/Vision/classification/image/swin_transformer 文件目录下
3. sh train_cifar.sh
4. 出现上述问题

**Comment by kaijieshi7** [comment]
> > 你试着 kill 一下僵尸进程呢？ 看样子虽然是 core dump 了，但是进程并没有退出 joy
> > 用这个命令找不到僵尸进程
> > ps -A -o stat,ppid,pid,cmd | grep -e '^[Zz]'
> > ![Screenshot from 2021-12-08 13-00-08](https://user-images.githubusercontent.com/19171201/145150950-e06f5649-c071-4a40-817c-c3d5d4189a31.png)

运行top命令，zombie=
![Screenshot from 2021-12-08 13-03-26](https://user-images.githubusercontent.com/19171201/145151271-c86479f6-7b0d-4bcc-b753-70febb06353d.png)
0

**Comment by chengtbf** [comment]
> > 你试着 kill 一下僵尸进程呢？ 看样子虽然是 core dump 了，但是进程并没有退出 joy
> > 用这个命令找不到僵尸进程
> > ps -A -o stat,ppid,pid,cmd | grep -e '^[Zz]'
> > ![Screenshot from 2021-12-08 13-00-08](https://user-images.githubusercontent.com/19171201/145150950-e06f5649-c071-4a40-817c-c3d5d4189a31.png)

确实。  gpu 显存是满的。但是并没有进程 hold 住这些显存。

**Comment by kaijieshi7** [comment]
好像是loss.backward()的问题。试了ADAMW和SGD，两个都一样不释放显存。
代码里面每行下面加了一个打印
![Screenshot from 2021-12-08 13-14-57](https://user-images.githubusercontent.com/19171201/145152274-b5f460b0-c4fe-4c79-ae79-563a57b99cee.png)

(torch_cuda10_1) kaijie@matrix:~/Documents/code/of/models/models/Vision/classification/image/swin_transformer$ sh train_cifar.sh 
+ OFRECORD_PATH=ofrecord
+ [ ! -d ofrecord ]
+ CHECKPOINT_PATH=checkpoints
+ [ ! -d checkpoints ]
+ LEARNING_RATE=0.0001
+ MOM=0.9
+ EPOCH=20
+ TRAIN_BATCH_SIZE=16
+ VAL_BATCH_SIZE=16
+ python3 train_cifar.py --save_checkpoint_path checkpoints --ofrecord_path ofrecord --learning_rate 0.0001 --mom 0.9 --epochs 20 --train_batch_size 16 --val_batch_size 16
libibverbs not available, ibv_fork_init skipped
Files already downloaded and verified
Files already downloaded and verified
init time : 0.8420534133911133
1
2
3
4
5
6
7
8
epoch 0 train iter 0 oneflow loss 6.994755268096924, train time : 0.42609548568725586
1
2
3
4
5
6
7
8
1
2
3
4
5
F1208 13:13:46.366336  2692 cuda_allocator.cpp:279] Check failed: piece != nullptr Error! : Out of memory when allocate size : 28901376
*** Check failure stack trace: ***
    @     0x7f65b06f856d  google::LogMessage::Fail()
    @     0x7f65b06f9dea  google::LogMessage::SendToLog()
    @     0x7f65b06f802d  google::LogMessage::Flush()
    @     0x7f65b06fb219  google::LogMessageFatal::~LogMessageFatal()
    @     0x7f65abc8f6c6  oneflow::vm::CudaAllocator::Allocate()
    @     0x7f65abcba6bd  oneflow::vm::ThreadSafeAllocator::Allocate()
    @     0x7f65aa869390  oneflow::vm::EagerBlobObject::TryAllocateBlobBodyMemory()
    @     0x7f65aa8c07ce  oneflow::vm::LocalCallOpKernelUtil::AllocateOutputBlobsMemory()
    @     0x7f65aa8c146a  oneflow::vm::LocalCallOpKernelUtil::Compute()
    @     0x7f65aa88dceb  oneflow::vm::LocalCallOpKernelInstructionType::Compute()
    @     0x7f65abc94752  oneflow::vm::CudaStreamType::Compute()
    @     0x7f65abcb374e  oneflow::vm::StreamType::Run()
    @     0x7f65abcc9225  oneflow::vm::VirtualMachineEngine::DispatchInstruction()
    @     0x7f65abccb68a  oneflow::vm::VirtualMachineEngine::DispatchAndPrescheduleInstructions()
    @     0x7f65abcbce58  oneflow::VirtualMachine::Loop()
    @     0x7f65b0da72df  (unknown)
    @     0x7f65ed9e36ba  start_thread
    @     0x7f65ed71951d  clone
    @              (nil)  (unknown)
Aborted (core dumped)

**Comment by chengtbf** [comment]
> 好像是loss.backward()的问题。试了ADAMW和SGD，两个都一样不释放显存。 

跟 Optimizer 无关。这个问题就是 后向计算申请内存 OOM 了，程序挂了。但是挂的时候没有清空显存。

这里有个帖子：

https://www.cnblogs.com/jins-note/p/11948927.html

Sometimes CUDA program crashed during execution, before memory was flushed. As a result, device memory remained occupied.

There are some solutions:

1.
Try using：

nvidia-smi --gpu-reset
or simply:
nvidia-smi -r
2.
Although it should be unecessary to do this in anything other than exceptional circumstances, the recommended way to do this on linux hosts is to unload the nvidia driver by doing

sudo rmmod nvidia 
with suitable root privileges and then reloading it with

sudo modprobe nvidia
If the machine is running X11, you will need to stop this manually beforehand, and restart it afterwards. The driver intialisation processes should eliminate any prior state on the device.

This answer has been assembled from comments and posted as a community wiki to get this question off the unanswered list for the CUDA tag

3.
This methods working for me:

check what is using your GPU memory with

sudo fuser -v /dev/nvidia*
 

Your output will look something like this:

复制代码
                     USER        PID  ACCESS COMMAND
/dev/nvidia0:        root       1256  F...m  Xorg
                     username   2057  F...m  compiz
                     username   2759  F...m  chrome
                     username   2777  F...m  chrome
                     username   20450 F...m  python
                     username   20699 F...m  python
复制代码
Then kill the PID that you no longer need on htop or with

sudo kill -9 PID.
4.
Or simply reboot:

sudo reboot

**Comment by lixinqi** [comment]
这个后续的解决办法是什么呢？

**Comment by chengtbf** [comment]
> 这个后续的解决办法是什么呢？

我马上提一个 PR。 OOM 之前 手动 reset 一下就行。

**Comment by chengtbf** [comment]
测试一下这个分支： dev_cc_cuda_mem_reset_in_oom
